src/kdf.rs

```diff
@@ -0,0 +1,351 @@
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_DERIVE,
+                },
+                mech_type: CKM_SP800_108_COUNTER_KDF,
+            }),
+        );
+        mechs.add_mechanism(
+            CKM_SP800_108_FEEDBACK_KDF,
+            Box::new(Sp800KDFMechanism {
+                info: CK_MECHANISM_INFO {
+                    ulMinKeySize: 0,
+                    ulMaxKeySize: std::u32::MAX as CK_ULONG,
+                    flags: CKF_DERIVE,
+                },
+                mech_type: CKM_SP800_108_FEEDBACK_KDF,
+            }),
+        );
+    }
+
+    fn parse_counter_format(
+        p: &CK_PRF_DATA_PARAM,
+    ) -> KResult<Sp800CounterFormat> {
+        if p.ulValueLen == 0 && p.pValue == std::ptr::null_mut() {
+            return Ok(Sp800CounterFormat {
+                defined: false,
+                le: false,
+                bits: 16,
+            });
+        }
+        if p.ulValueLen as usize
+            != ::std::mem::size_of::<CK_SP800_108_COUNTER_FORMAT>()
+        {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+        let cf = unsafe { *(p.pValue as *const CK_SP800_108_COUNTER_FORMAT) };
+        Ok(Sp800CounterFormat {
+            defined: true,
+            le: match cf.bLittleEndian {
+                0 => false,
+                1 => true,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+            },
+            bits: match cf.ulWidthInBits {
+                8 | 16 | 24 | 32 => cf.ulWidthInBits as usize,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+            },
+        })
+    }
+
+    fn parse_byte_array(p: &CK_PRF_DATA_PARAM) -> KResult<Vec<u8>> {
+        if p.ulValueLen == 0 {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+        Ok(bytes_to_vec!(p.pValue, p.ulValueLen))
+    }
+
+    fn parse_dkm_length(
+        p: &CK_PRF_DATA_PARAM,
+    ) -> KResult<Sp800DKMLengthFormat> {
+        if p.ulValueLen as usize
+            != ::std::mem::size_of::<CK_SP800_108_DKM_LENGTH_FORMAT>()
+        {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+        let dkm =
+            unsafe { *(p.pValue as *const CK_SP800_108_DKM_LENGTH_FORMAT) };
+        Ok(Sp800DKMLengthFormat {
+            method: match dkm.dkmLengthMethod {
+                CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS
+                | CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS => {
+                    dkm.dkmLengthMethod
+                }
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+            },
+            le: match dkm.bLittleEndian {
+                0 => false,
+                1 => true,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+            },
+            bits: match dkm.ulWidthInBits {
+                8 | 16 | 24 | 32 => dkm.ulWidthInBits as usize,
+                _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
```

Comment:
Same comment for counter... our encoded length has to be 32 bits or less. That doesn't mean length can't be more, just that longer lengths will wrap (which is now this is defined anyway). I don't know if SP 800 specifies that lengths can't be greater than 32 bits. NSS max is 64 bits because thats how big the underlying counter is. I don't know if any real world use would actually have more than 32 bits here.

Reply:
Uhm this one here is kinda wrong, I need to make it go up to 64bit, just because the spec puts no limits on it, but probably nothing will ever come close to 32 bit even ... yet there is no point to restrict to 32 bit, 64bit values may be used in some KDF that wants to take in a 64bit usize/size_t parameters for lengths.